Add prefix‑aware ACL permission checks and new module API

eifrah-aws opened the pull request:
## Description

This change introduces the ability for modules to check ACL permissions against key prefix. The update adds a dedicated `prefixmatchlen` helper and extends the core ACL selector logic to support a prefix‑matching mode.

The new API `ValkeyModule_ACLCheckPrefixPermissions` is registered and exposed to modules, and a corresponding implementation is added in `module.c`. Existing internal callers that already perform prefix checks (e.g., `VM_ACLCheckKeyPermissions`) are updated to use the new flag, while all legacy paths remain unchanged.

The change also modifies the `aclcheck§ test module that exercises the new prefix‑checking API, ensuring that read/write operations are correctly allowed or denied based on the ACL configuration.

Key areas touched:

* ACL logic
* Module API
* Testing

# Motivation

The search module presently makes costly calls to verify index permissions (see https://github.com/valkey-io/valkey-search/blob/main/src/acl.cc#L295). This PR introduces a more efficient approach for that.

eifrah-aws committed:
Add prefix‑aware ACL permission checks and new module API

This change introduces the ability for modules to check ACL permissions against key prefixe. The update adds a dedicated `prefixmatchlen` helper and extends the core ACL selector logic to support a prefix‑matching mode.

The new API `ValkeyModule_ACLCheckKeyPrefixPermissions` is registered and exposed to modules, and a corresponding implementation is added in `module.c`. Existing internal callers that already perform prefix checks (e.g., `VM_ACLCheckKeyPermissions`) are updated to use the new flag, while all legacy paths remain unchanged.

The change also adds a small test module that exercises the new prefix‑checking API, ensuring that read/write operations are correctly allowed or denied based on the ACL configuration.

Key areas touched:

* ACL logic
* Module API
* Testing

Signed-off-by: Eran Ifrah <eifrah@amazon.com>

rjd15372 commented on src/util.c:
```diff
@@ -196,6 +196,20 @@ int stringmatch(const char *pattern, const char *string, int nocase) {
+    } else {
+        // Call existing string match algorithm
+        int skipLongerMatches = 0;
+        return stringmatchlen_impl(pattern, patternLen, string, stringLen, nocase, &skipLongerMatches, 0);
```
Can't this be replaced by 
```c
stringmatchlen(pattern, patternLen, string, stringLen, nocase);
```

eifrah-aws replied:
Correct, I missed that :)

eifrah-aws replied:
Done.

ranshid commented on src/acl.c:
```diff
@@ -1541,11 +1541,26 @@ user *ACLGetUserByName(const char *name, size_t namelen) {
+ * treated as a prefix. If the selector has the {@code SELECTOR_FLAG_ALLKEYS} flag set,
+ * the check always succeeds.
+ *
+ * @param selector      The ACL selector containing the allowed key patterns.
```
We do not make use of such tags in our code base (aside dependancies that already include them).
Even if I like presenting doxigen parsable documentation, I would hate introducing it without apply a clear development guidlines to do so and maintainers acceptance about changing the format.

eifrah-aws replied:
Fixed.

ranshid commented on src/acl.c:
```diff
@@ -1564,6 +1579,12 @@ static int ACLSelectorCheckKey(aclSelector *selector, const char *key, int keyle
         keyPattern *pattern = listNodeValue(ln);
         if ((pattern->flags & key_flags) != key_flags) continue;
         size_t plen = sdslen(pattern->pattern);
+        if (is_prefix) {
+            if (prefixmatchlen(pattern->pattern, plen, key, keylen, 0))
+                return ACL_OK;
+            else
+                continue;
+        }
```
I think there is much shared logic that makes it reasonable to add is_prefix as argument.
But maybe it will be clearer if we turn this into a static internal function and introduce external ACLSelectorCheckKey ACLSelectorCheckKeyPrefix which will call the internal one?

eifrah-aws replied:
This function is already `static` and used by other methods

ranshid commented on src/acl.c:
```diff
@@ -1721,13 +1742,29 @@ static int ACLSelectorCheckCmd(aclSelector *selector,
- * the ACLs associated with the specified user according to the
- * keyspec access flags.
+/**
+ * @brief Checks whether the given user has permission to access a specified key.
```
Again, lets remove all the special documentation format tags.

allenss-amazon commented:
This addresses issue https://github.com/valkey-io/valkey/issues/2764

madolson commented on src/util.c:
```diff
@@ -196,6 +196,19 @@ int stringmatch(const char *pattern, const char *string, int nocase) {
+int prefixmatchlen(const char *pattern, int patternLen, const char *string, int stringLen, int nocase) {
+    if (stringLen == 0) {
+        // Check if is an all access pattern
+        return patternLen == 1 && pattern[0] == '*';
```
Is this a required optimization? Prefixmatchlen should not be called with the `SELECTOR_FLAG_ALLKEYS` flag set.

eifrah-aws replied:
I don't see harm in leaving this code here, do you?

madolson replied:
If it's just an optimization, just make it clear from the comment.

eifrah-aws replied:
Done. I also added a TCL test to check for "*" patterns.

madolson commented on src/valkeymodule.h:
```diff
@@ -2159,9 +2159,12 @@ VALKEYMODULE_API void (*ValkeyModule_ScriptingEngineDebuggerFlushLogs)(void) VAL
 VALKEYMODULE_API void (*ValkeyModule_ScriptingEngineDebuggerProcessCommands)(int *client_disconnected,
                                                                              ValkeyModuleString **err) VALKEYMODULE_ATTR;
 
+VALKEYMODULE_API int (*ValkeyModule_ACLCheckPrefixPermissions)(ValkeyModuleCtx *ctx,
```
```suggestion
VALKEYMODULE_API int (*ValkeyModule_ACLCheckKeyPrefixPermissions)(ValkeyModuleCtx *ctx,
```
Minor preference to include key in the name.

eifrah-aws replied:
Done

madolson commented on src/acl.c:
```diff
@@ -1721,13 +1742,31 @@ static int ACLSelectorCheckCmd(aclSelector *selector,
-int ACLUserCheckKeyPerm(user *u, const char *key, int keylen, int flags) {
+ * This function verifies the access control list (ACL) permissions for a user on a key.
+ * If the user pointer is  NULL, the connection is considered authorized to run any
+ * command and the function returns  ACL_OK.  Otherwise, the function iterates
```
```suggestion
 * command and the function returns ACL_OK. Otherwise, the function iterates
```
There are a bunch of double spaces in the comments you added, just double check if they were intended.

madolson commented on src/acl.c:
```diff
@@ -1541,11 +1541,26 @@ user *ACLGetUserByName(const char *name, size_t namelen) {
+ * - keyspec_flags Bitmask of requested key operations (e.g.,  CMD_KEY_ACCESS,
+ *                 CMD_KEY_INSERT,  CMD_KEY_DELETE,  CMD_KEY_UPDATE).
+ * - is_prefix     If non-zero, the key is compared as a prefix against the selector's
+ *                 patterns using  prefixmatchlen rather than a full string match.
```
Now that we use bools, this seems like it should be a bool and not an int.

madolson commented on src/module.c:
```diff
@@ -14136,6 +14136,26 @@ int VM_GetDbIdFromDefragCtx(ValkeyModuleDefragCtx *ctx) {
 
+/**
+ * To check if user has access to the keyspace with the given index prefix
+ */
```
This comment get's rendered in to https://valkey.io/topics/modules-api-ref/, so it should be as descriptive as the other ACLCheck APIs.

eifrah-aws replied:
Updated to a more comprehensive doc

madolson commented on src/module.c:
```diff
@@ -14136,6 +14136,26 @@ int VM_GetDbIdFromDefragCtx(ValkeyModuleDefragCtx *ctx) {
 }
 
+/**
+ * To check if user has access to the keyspace with the given index prefix
```
This API uses the current contexts user, which is not mirrored from the other APIs like `ACLCheckKeyPermissions`.

eifrah-aws replied:
Solved now sorry for missing it.

eifrah-aws committed:
Changed comment style for ACL verification function

Signed-off-by: Eran Ifrah <eifrah@amazon.com>

ranshid commented on src/acl.c:
```diff
@@ -1541,11 +1541,26 @@ user *ACLGetUserByName(const char *name, size_t namelen) {
 
-/* Check if the key can be accessed by the selector.
+/**
+ * Checks whether a given key is allowed for the specified ACL selector.
```
```suggestion
/* Checks whether a given key is allowed for the specified ACL selector.
```

ranshid commented on src/acl.c:
```diff
@@ -1721,13 +1742,31 @@ static int ACLSelectorCheckCmd(aclSelector *selector,
- * the ACLs associated with the specified user according to the
- * keyspec access flags.
+/**
+ * Checks whether the given user has permission to access a specified key.
```
```suggestion
/* Checks whether the given user has permission to access a specified key.
```

ranshid commented on src/util.c:
```diff
@@ -196,6 +196,20 @@ int stringmatch(const char *pattern, const char *string, int nocase) {
+int prefixmatchlen(const char *pattern, int patternLen, const char *string, int stringLen, int nocase) {
+    if (patternLen == 1 && pattern[0] == '*') {
+        // Minor optimization: fast path: avoid calling "stringmatchlen" if the input pattern is exactly "*":
+        // We always return 1 in this case.
```
```suggestion
        /* Minor optimization: fast path: avoid calling "stringmatchlen" if the input pattern is exactly "*":
         * We always return 1 in this case. */
```

ranshid reviewed (approved):
LGTM. 
Added some comment nits.

@madolson I think all was addressed, I think we can merge it.

madolson commented on src/module.c:
```diff
@@ -14136,6 +14136,38 @@ int VM_GetDbIdFromDefragCtx(ValkeyModuleDefragCtx *ctx) {
+ *
+ *  Otherwise it returns `VALKEYMODULE_OK`.
+ */
+int VM_ACLCheckKeyPrefixPermissions(ValkeyModuleCtx *ctx, const char *key, size_t len, unsigned int flags) {
```
I made an earlier comment, why is this on the context and not the user, which mirrors the other APIs? I would rather mirror the other APIs.

eifrah-aws replied:
I missed it (replied on that comment) and it is now fixed. Change the signature to use `ValkeyModuleUser` instead.

eifrah-aws committed:
Change ACLCheckKeyPrefixPermissions To Accept ValkeyModuleUser Instead Of Context

The API signature of `ValkeyModule_ACLCheckKeyPrefixPermissions` has been updated to accept a `ValkeyModuleUser*` parameter instead of `ValkeyModuleCtx*`.

Signed-off-by: Eran Ifrah <eifrah@amazon.com>

ranshid commented on src/module.c:
```diff
@@ -14159,6 +14159,43 @@ int VM_GetDbIdFromDefragCtx(ValkeyModuleDefragCtx *ctx) {
+int VM_ACLCheckKeyPrefixPermissions(ValkeyModuleUser *user, const char *key, size_t len, unsigned int flags) {
+    const int allow_mask = (VALKEYMODULE_CMD_KEY_ACCESS | VALKEYMODULE_CMD_KEY_INSERT | VALKEYMODULE_CMD_KEY_DELETE | VALKEYMODULE_CMD_KEY_UPDATE);
+
+    if (user == NULL || user->user == NULL) {
```
agree that user should be provided and we should not accept null.
the 2nd check, though is contradicting to the way we operate this logic. when no user the module can do everything. We should just pass it on and have ACLUserCheckKeyPerm as is.

eifrah-aws replied:
Fixed.

eifrah-aws committed:
Addressed more comments.

Signed-off-by: Eran Ifrah <eifrah@amazon.com>

eifrah-aws committed:
Addressed more comments.

Signed-off-by: Eran Ifrah <eifrah@amazon.com>

madolson committed:
Update src/util.c

Co-authored-by: Ran Shidlansik <ranshid@amazon.com>
Signed-off-by: Madelyn Olson <madelyneolson@gmail.com>

madolson reviewed (approved):
LGTM: Will a second person from @valkey-io/core-team give this a thumbs up?

enjoy-binbin commented on src/acl.c:
```diff
@@ -1690,7 +1711,7 @@ static int ACLSelectorCheckCmd(aclSelector *selector,
         for (int j = 0; j < result->numkeys; j++) {
             int idx = resultidx[j].pos;
-            ret = ACLSelectorCheckKey(selector, argv[idx]->ptr, sdslen(argv[idx]->ptr), resultidx[j].flags);
+            ret = ACLSelectorCheckKey(selector, argv[idx]->ptr, sdslen(argv[idx]->ptr), resultidx[j].flags, 0);
```
```suggestion
            ret = ACLSelectorCheckKey(selector, argv[idx]->ptr, sdslen(argv[idx]->ptr), resultidx[j].flags, false);
```

enjoy-binbin commented on src/util.c:
```diff
@@ -196,6 +196,20 @@ int stringmatch(const char *pattern, const char *string, int nocase) {
+         * We always return 1 in this case. */
+        return 1;
+    } else if (patternLen > 0 && pattern[patternLen - 1] != '*') {
+        // Reject the pattern if it doesn't end with '*'
```
```suggestion
        /* Reject the pattern if it doesn't end with '*' */
```

enjoy-binbin commented on src/util.c:
```diff
@@ -196,6 +196,20 @@ int stringmatch(const char *pattern, const char *string, int nocase) {
+        // Reject the pattern if it doesn't end with '*'
+        return 0;
+    } else {
+        // Call existing string match algorithm
```
```suggestion
        /* Call existing string match algorithm */
```

enjoy-binbin reviewed (approved):
LGTM.

ranshid committed:
Apply suggestions from code review

apply PR comments

Co-authored-by: Binbin <binloveplay1314@qq.com>
Signed-off-by: Ran Shidlansik <ranshid@amazon.com>
